experimental/sql/dialects/rel_tree.py

```diff
@@ -0,0 +1,223 @@
+  """
+  name = "rel_tree.int32"
+
+
+@irdl_attr_definition
+class String(DataType):
+  """
+  Models a string type in a relational query, that can be either nullable or
+  not.
+
+  Example:
+
+  ```
+  !rel_tree.string<0: !i1>
+  ```
+  """
+  name = "rel_tree.string"
+
+  nullable = ParameterDef(IntegerAttr)
+
+  @staticmethod
+  @builder
+  def get(val: int) -> 'String':
+    return String([IntegerAttr.from_int_and_width(val, 1)])
+
+
+@irdl_op_definition
+class Column(Operation):
+  """
+  References a specific column with name `col_name`.
+
+  Example:
+
+  ```
+  rel_tree.column() ["col_name" = "a"]
+  ```
+  """
+  name = "rel_tree.column"
+
+  col_name = AttributeDef(StringAttr)
+
+  @builder
+  @staticmethod
+  def get(table: Region, col_name: str) -> 'Column':
+    return Column.build(attributes={"col_name": StringAttr.from_str(col_name)},
+                        regions=[table])
+
+
+@irdl_op_definition
+class Predicate(Operation):
+  """
+  Applies the `comparator` to `left` and `right`.
+
+  Example:
+
+  ```
+  rel_tree.predicate() ["comparator" = "="] {
+    rel_tree.column() ...
+  } {
+    rel_tree.literal() ...
+  }
+  ```
+  """
+  name = "rel_tree.predicate"
+
+  comparator = AttributeDef(StringAttr)
+  left = SingleBlockRegionDef()
+  right = SingleBlockRegionDef()
+
+  @builder
+  @staticmethod
+  def get(comparator: str, left: Region, right: Region) -> 'Predicate':
+    return Predicate.build(
+        attributes={"comparator": StringAttr.from_str(comparator)},
+        regions=[left, right])
+
+
+@irdl_op_definition
+class Selection(Operation):
+  """
+  Selects all tuples from `table` that fulfill `predicates`.
+
+  Example:
+
+  ```
+  rel_tree.selection() {
+    rel_tree.pandas_table() ...
+  } {
+    rel_tree.predicate() ...
+  }
+  ```
+  """
+  name = "rel_tree.selection"
+
+  table = SingleBlockRegionDef()
+  predicates = SingleBlockRegionDef()
+
+  @staticmethod
+  @builder
+  def get(table: Region, predicates: Region) -> 'Selection':
+    return Selection.build(regions=[table, predicates])
+
+
+@irdl_op_definition
+class PandasTable(Operation):
+  """
+  Defines a table with name `table_name` and schema `schema`.
+
+  Example:
+
+  ```
+  rel_tree.pandas_table() ["table_name" = "t"] {
+    rel_tree.schema_element() ...
+    ...
+  }
+  ```
+  """
+  name = "rel_tree.pandas_table"
+
+  table_name = AttributeDef(StringAttr)
+  schema = SingleBlockRegionDef()
+
+  @staticmethod
+  @builder
+  def get(name: str, Schema: Region) -> 'PandasTable':
+    return PandasTable.build(
+        attributes={"table_name": StringAttr.from_str(name)}, regions=[Schema])
+
+
+@irdl_op_definition
+class SchemaElement(Operation):
+  """
+  Defines a schema element with name `elt_name` and type `elt_type`.
+
+  Example:
+
+  ```
+  rel_tree.schema_element() ["elt_name" = "id", "elt_type" = !rel_tree.int32]
+  ```
+  """
+  name = "rel_tree.schema_element"
+
+  elt_name = AttributeDef(StringAttr)
+  elt_type = AttributeDef(DataType)
```

Comment:
These *could* also be just `name` and `type`... Although you'd clash with the keyword `type`, right? Maybe this is a good enough work-around then...

Reply:
name actually clashes as well, since xDSL expects the string that gets parsed to a certain op to be a field named name (i.e. the field on line 174).